utils/log.go

```diff
@@ -0,0 +1,96 @@
+package utils
+
+import "fmt"
+
+func FormatWritesLog(codes []byte, line *string) {
+	n := len(codes)
+	str := ""
+	exist := false
+	if n == 3 {
+		if str, exist = ASCIIGroupToStr(fmt.Sprintf("%X", codes)); exist {
+			*line += str
+			codes = nil
+		}
+	}
+	// sh control codes
+	if n >= 6 {
+		if str, exist = ASCIIGroupToStr(fmt.Sprintf("%X", []byte{codes[0], codes[1], codes[n-1]})); exist {
+			*line += str
+			codes = nil
+		}
+	}
+
+	if codes != nil {
+		str = ASCIIToStr(codes)
+		*line += str
+	}
+
+	return
+}
+
+func ASCIIToStr(codes []byte) string {
+	control := map[byte]string{
+		0:   "NUL",
+		1:   "SOH",
+		2:   "STX",
+		3:   "ETX",
+		4:   "EOT",
+		5:   "ENQ",
+		6:   "ACK",
+		7:   "BEL",
+		8:   "BS",
+		9:   "HT",
+		10:  "LF",
+		11:  "VT",
+		12:  "FF",
+		13:  "CR",
+		14:  "SO",
+		15:  "SI",
+		16:  "DLE",
+		17:  "DCI",
+		18:  "DC2",
+		19:  "DC3",
+		20:  "DC4",
+		21:  "NAK",
+		22:  "SYN",
+		23:  "TB",
+		24:  "CAN",
+		25:  "EM",
+		26:  "SUB",
+		27:  "ESC",
+		28:  "FS",
+		29:  "GS",
+		30:  "RS",
+		31:  "US",
+		32:  "SPACE",
+		127: "DEL",
+	}
```

Comment:
This should be defined outside the function. It never changes, so there's no need to create a fresh one each time we call `ASCIIToStr`.